Realm/Tests/RealmTests.mm

```diff
@@ -1455,6 +1455,438 @@ - (void)testActiveVersionLimit {
+    
+    [realm beginWriteTransaction];
+    [realm createObject:StringObject.className withValue:@[@"string 2"]];
+    [realm commitAsyncWriteTransaction:^(NSError *error) {
+        [asyncComplete fulfill];
+    }];
+
+    [self waitForExpectationsWithTimeout:1.0 handler:nil];
+
+    XCTAssertEqual(2, [StringObject allObjectsInRealm:realm].count);
+}
+
+- (void)testAsyncTransactionMixedWithSync {
+    RLMRealm *realm = RLMRealm.defaultRealm;
+    XCTestExpectation *asyncComplete = [self expectationWithDescription:@"async transaction complete"];
+
+    XCTAssertEqual(0, [StringObject allObjectsInRealm:realm].count);
+
+    [realm beginAsyncWriteTransaction:^{
+        [realm createObject:StringObject.className withValue:@[@"string"]];
+        [realm commitAsyncWriteTransaction];
+    }];
+
+    [realm beginAsyncWriteTransaction:^{
+        [realm createObject:StringObject.className withValue:@[@"string 2"]];
+        [realm commitAsyncWriteTransaction:^(NSError *error) {
+            [asyncComplete fulfill];
+        }];
+    }];
+
+    [realm beginWriteTransaction];
+    [realm createObject:StringObject.className withValue:@[@"string 3"]];
+    [realm commitWriteTransaction];
+
+    [self waitForExpectationsWithTimeout:1.0 handler:nil];
+
+    XCTAssertEqual(3, [StringObject allObjectsInRealm:realm].count);
+}
+
+- (void)testAsyncTransactionMixedWithCancelledSync {
+    RLMRealm *realm = RLMRealm.defaultRealm;
+    XCTestExpectation *asyncComplete = [self expectationWithDescription:@"async transaction complete"];
+
+    XCTAssertEqual(0, [StringObject allObjectsInRealm:realm].count);
+
+    [realm beginAsyncWriteTransaction:^{
+        [realm createObject:StringObject.className withValue:@[@"string"]];
+        [realm commitAsyncWriteTransaction];
+    }];
+
+    [realm beginAsyncWriteTransaction:^{
+        [realm createObject:StringObject.className withValue:@[@"string 2"]];
+        [realm commitAsyncWriteTransaction:^(NSError *error) {
+            [asyncComplete fulfill];
+        }];
+    }];
+
+    [realm beginWriteTransaction];
+    [realm createObject:StringObject.className withValue:@[@"string 3"]];
+    [realm cancelWriteTransaction];
+
+    [self waitForExpectationsWithTimeout:1.0 handler:nil];
+
+    XCTAssertEqual(2, [StringObject allObjectsInRealm:realm].count);
+}
+
+- (void)testAsyncTransactionChangeNotification {
+    RLMRealm *realm = RLMRealm.defaultRealm;
+    __block XCTestExpectation *asyncComplete = [self expectationWithDescription:@"async transaction complete"];
+    asyncComplete.expectedFulfillmentCount = 4;
+
+    RLMResults<RLMObject *> *resultsUnderTest = [StringObject allObjects];
+    RLMNotificationToken *token = [resultsUnderTest addNotificationBlock:^(RLMResults<RLMObject *> * _Nullable results, RLMCollectionChange * _Nullable change, NSError *error) {
+        if (!change)
+            return; // ignore initial
+        XCTAssertNotNil(change);
+        [asyncComplete fulfill];
```

Comment:
There's a race condition here. The first (or both) writes can happen before the initial notification, which results in fewer than four total calls to `fulfill`. You have to wait for the initial notification to be delivered before performing the writes.